### PR TITLE
feat: add platform flag to cosign copy command

### PR DESCRIPTION
**Pull request by vishal-chdhry**
Signed-off-by: Vishal Choudhary <sendtovishalchoudhary@gmail.com>

#### Summary

Fixes: #2748

This PR adds a `--platform` flag to `cosign copy`. This allows copying images and signatures for specific platform/architecture.

/cc @imjasonh @znewman01 @cpanato

Before:
```bash
cosign copy alpine ttl.sh/test-cosign-copy:5m
Copying index.docker.io/library/alpine@sha256:5febc00b4d2a84af2a077bc34ea90659b6570110a54253f19c5dca8164b1dbf6 to ttl.sh/test-cosign-copy:sha256:5febc00b4d2a84af2a077bc34ea90659b6570110a54253f19c5dca8164b1dbf6...
Copying index.docker.io/library/alpine@sha256:b312e4b0e2c665d634602411fcb7c2699ba748c36f59324457bc17de485f36f6 to ttl.sh/test-cosign-copy:sha256:b312e4b0e2c665d634602411fcb7c2699ba748c36f59324457bc17de485f36f6...
Copying index.docker.io/library/alpine@sha256:f748290eb66ad6f938e25dd348acfb3527a422e280b7547b1cdfaf38d4492c4b to ttl.sh/test-cosign-copy:sha256:f748290eb66ad6f938e25dd348acfb3527a422e280b7547b1cdfaf38d4492c4b...
Copying index.docker.io/library/alpine@sha256:16e86b2388774982fbdf230101a72201691b1f97cb0066c2099abf30dd7e6d59 to ttl.sh/test-cosign-copy:sha256:16e86b2388774982fbdf230101a72201691b1f97cb0066c2099abf30dd7e6d59...
Copying index.docker.io/library/alpine@sha256:c75ede79e457d6454bca6fc51967a247a4b9daff9f31197cfbef69b1a651cada to ttl.sh/test-cosign-copy:sha256:c75ede79e457d6454bca6fc51967a247a4b9daff9f31197cfbef69b1a651cada...
Copying index.docker.io/library/alpine@sha256:7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a to ttl.sh/test-cosign-copy:sha256:7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a...
Copying index.docker.io/library/alpine@sha256:c5c5fda71656f28e49ac9c5416b3643eaa6a108a8093151d6d1afc9463be8e33 to ttl.sh/test-cosign-copy:sha256:c5c5fda71656f28e49ac9c5416b3643eaa6a108a8093151d6d1afc9463be8e33...
Copying index.docker.io/library/alpine@sha256:1fd62556954250bac80d601a196bb7fd480ceba7c10e94dd8fd4c6d1c08783d5 to ttl.sh/test-cosign-copy:sha256:1fd62556954250bac80d601a196bb7fd480ceba7c10e94dd8fd4c6d1c08783d5...
Copying index.docker.io/library/alpine@sha256:7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a to ttl.sh/test-cosign-copy:5m...
```

After:
```bash
go run ./cmd/cosign copy --platform=linux/amd64 alpine ttl.sh/test-cosign-copy-local:5m
Copying index.docker.io/library/alpine@sha256:c5c5fda71656f28e49ac9c5416b3643eaa6a108a8093151d6d1afc9463be8e33 to ttl.sh/test-cosign-copy-local:sha256-c5c5fda71656f28e49ac9c5416b3643eaa6a108a8093151d6d1afc9463be8e33...
Copying index.docker.io/library/alpine@sha256:c5c5fda71656f28e49ac9c5416b3643eaa6a108a8093151d6d1afc9463be8e33 to ttl.sh/test-cosign-copy-local:5m...
```
#### Release Note
Added `--platform` flag to `cosign copy` command

### CODE REVIEW

File: cmd/cosign/cli/common/platform.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package common
```
Comment (znewman01) [inline cmd/cosign/cli/common/platform.go:1]: I think I'd prefer to see this in something like `pkg/oci`

File: cmd/cosign/cli/copy/copy_test.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ func TestCopyAttachmentTagPrefix(t *testing.T) {
 
 	err := CopyCmd(ctx, options.RegistryOptions{
 		RefOpts: refOpts,
-	}, srcImg, destImg, false, true)
```
Comment (znewman01) [inline cmd/cosign/cli/copy/copy_test.go:36]: I'd like to see a test that exercises the new behavior.

File: cmd/cosign/cli/copy/copy.go [hunk 3]
```diff
@@ -71,6 +72,43 @@ func CopyCmd(ctx context.Context, regOpts options.RegistryOptions, srcImg, dstIm
 		return err
 	}
 
+	idx, isIndex := root.(oci.SignedImageIndex)
+
```
Comment (znewman01) [inline cmd/cosign/cli/copy/copy.go:76]: I know a lot of this is copy-paste from `sbom download`. Is it possible to factor out?

File: cmd/cosign/cli/copy/copy.go [hunk 4]
```diff
@@ -71,6 +72,43 @@ func CopyCmd(ctx context.Context, regOpts options.RegistryOptions, srcImg, dstIm
...
+			return fmt.Errorf("parsing platform: %w", err)
+		}
+		platforms, err := common.GetIndexPlatforms(idx)
+		if err != nil {
+			return fmt.Errorf("getting available platforms: %w", err)
+		}
+
+		platforms = common.MatchPlatform(targetPlatform, platforms)
+		if len(platforms) == 0 {
+			return fmt.Errorf("unable to find an SBOM for %s", targetPlatform.String())
```
Comment (znewman01) [inline cmd/cosign/cli/copy/copy.go:93]: bad message after copy/paste 🙂

File: cmd/cosign/cli/copy/copy.go [hunk 5]
```diff
@@ -71,6 +72,43 @@ func CopyCmd(ctx context.Context, regOpts options.RegistryOptions, srcImg, dstIm
 		return err
 	}
 
+	idx, isIndex := root.(oci.SignedImageIndex)
+
+	if platform != "" && !isIndex {
```
Comment (znewman01) [inline cmd/cosign/cli/copy/copy.go:77]: IMO this is clearer as:

```
func RootForPlatform(root, platform string) string, error {
    if platform == "" {
        // Copy all platforms
        return root, nil
    }
    if !isIndex {
        return fmt.Errorf("...")
    }
   // ...
}

Comment (znewman01) [reply cmd/cosign/cli/copy/copy.go:77]: That code won't compile as-is, there's a couple other things to thread through but I hope it gets the idea across.

**Review by znewman01 — CHANGES_REQUESTED** [review]
Thank you very much for this change! I left some minor feedback

**Comment by vishal-chdhry** [comment]
@znewman01 I have added a unit test as well as a example in pull request description. Can you please review it again

File: doc/cosign_copy.md [hunk 6]
```diff
@@ -19,6 +19,9 @@ cosign copy [flags]
 
   # overwrite destination image and signatures
   cosign copy -f example.com/src example.com/dest
+
+	# copy a container image and its signatures for a specific platform
```
Comment (znewman01) [inline doc/cosign_copy.md:23]: nitpick: can you dedent this line?

**Review by znewman01 — APPROVED** [review]

**Pull request merged**